alpa/pipeline_parallel/cross_mesh_resharding.py

```diff
@@ -1140,6 +1228,231 @@ def _generate_send_recv_resharding_strategy_by_loads(
+
+            for src_tileslice_idx, src_tileslice in enumerate(src_tileslices):
+                per_spec_plan[
+                    src_tileslice_idx] = src_tileslice.replica_device_strs[0]
+            per_spec_plans.append(per_spec_plan)
+        strategy = ReshardingStrategy(per_spec_plans, None)
+        return strategy
+
+    @staticmethod
+    def _generate_broadcast_resharding_strategy_by_loadbalance(
+            spec, src_mesh, dst_mesh):
+        """
+            Generate the broadcast-based resharding strategy by balancing
+            loads and along time.
+        """
+
+        def device_id_to_str(host_ip, device_id, device_type="gpu"):
+            """Convert device id (int) to a canonical device string."""
+            return f"{host_ip}:{device_type}:{device_id}"
+
+        def get_resources_info_in_mesh(mesh):
+            device_strs = []
+            device_host_map = {}
+            nic_constraints = []
+
+            for i in range(mesh.num_hosts):
+                ip = mesh.host_info[i]["NodeManagerAddress"]
+                one_nic_constraint = []
+                for device in mesh.devices[i]:
+                    device_str = device_id_to_str(ip, device)
+                    device_strs.append(device_str)
+                    one_nic_constraint.append(device_str)
+                    # Here I assume there is only one NIC in one host.
+                    device_host_map[device_str] = ip
+                nic_constraints.append(one_nic_constraint)
+            return device_strs, device_host_map, nic_constraints
+
+        src_mesh_devices, src_device_host_map, src_nic_constraints = (
+            get_resources_info_in_mesh(src_mesh))
+
+        dst_mesh_devices, dst_device_host_map, dst_nic_constraints = (
+            get_resources_info_in_mesh(dst_mesh))
+        device_host_map = {**src_device_host_map, **dst_device_host_map}
+        nic_constraints = src_nic_constraints + dst_nic_constraints
+
+        works = []
+        for i, (dst_tile, src_tileslices,
+                _) in enumerate(spec.dst_tile_to_src_tiles_map):
+            for j, src_tileslice in enumerate(src_tileslices):
+                senders = src_tileslice.replica_device_strs
+                receivers = dst_tile.replica_device_strs
+                data_size = src_tileslice.tile_size
+                works.append(
+                    SingleReshardingLoadBalancingWork(senders, receivers,
+                                                      data_size))
+
+        task = ReshardingLoadBalancingTaskSolver(src_mesh_devices,
+                                                 dst_mesh_devices,
+                                                 device_host_map, works,
+                                                 nic_constraints)
+
+        sol_assigned_sender, sol_order = task.solve()
+
+        per_spec_plans = []
+        rank_to_idx = {}
+        cnt = 0
+        for i, (dst_tile, src_tileslices,
```

Comment:
this post-processing seems similar with what in `_generate_send_recv_resharding_strategy_by_loadbalance`, can we use a function to let them share the code?

Reply:
Yep, they are indeed similar, but It seems not so easy to merge them together. They have different loop structures(2 for loops vs 3 for loops) and different pieces of code.

Reply:
Right. I misread it. Could you please just add a comment line to separate the post-process and other codes? Besides, seems like you can use a list to replace the `rank_to_idx`, and use `.append` instead of `[cnt] = ...`+`cnt += 1`(the same happens for `send_recv_resharding_strategy`)?

Reply:
OK